st2common/st2common/util/mongoescape.py

```diff
@@ -19,70 +19,62 @@
 
 from st2common.util.ujson import fast_deepcopy
 
+# Note: Because of old rule escaping code, two different characters can be translated back to dot
+RULE_CRITERIA_UNESCAPED = ['.']
+RULE_CRITERIA_ESCAPED = [u'\u2024']
+RULE_CRITERIA_ESCAPE_TRANSLATION = dict(list(zip(RULE_CRITERIA_UNESCAPED, RULE_CRITERIA_ESCAPED)))
+RULE_CRITERIA_UNESCAPE_TRANSLATION = dict(list(zip(RULE_CRITERIA_ESCAPED, RULE_CRITERIA_UNESCAPED)))
+
 # http://docs.mongodb.org/manual/faq/developers/#faq-dollar-sign-escaping
 UNESCAPED = ['.', '$']
 ESCAPED = [u'\uFF0E', u'\uFF04']
 ESCAPE_TRANSLATION = dict(list(zip(UNESCAPED, ESCAPED)))
-UNESCAPE_TRANSLATION = dict(list(zip(ESCAPED, UNESCAPED)))
+UNESCAPE_TRANSLATION = dict(
+    list(zip(ESCAPED, UNESCAPED)) + list(zip(RULE_CRITERIA_ESCAPED, RULE_CRITERIA_UNESCAPED))
+)
 
-# Note: Because of old rule escaping code, two different characters can be translated back to dot
-RULE_CRITERIA_UNESCAPED = ['.']
-RULE_CRITERIA_ESCAPED = [u'\u2024']
-RULE_CRITERIA_ESCAPE_TRANSLATION = dict(list(zip(RULE_CRITERIA_UNESCAPED,
-                                            RULE_CRITERIA_ESCAPED)))
-RULE_CRITERIA_UNESCAPE_TRANSLATION = dict(list(zip(RULE_CRITERIA_ESCAPED,
-                                              RULE_CRITERIA_UNESCAPED)))
 
+def _translate_chars(field, translation):
+    if isinstance(field, list):
+        return _translate_chars_in_list(field, translation)
 
-def _prep_work_items(d):
-    return [(k, v, d) for k, v in six.iteritems(d)]
+    if isinstance(field, dict):
+        return _translate_chars_in_dict(field, translation)
 
+    return field
 
-def _translate_chars(field, translation):
-    # Only translate the fields of a dict
-    if not isinstance(field, dict):
-        return field
 
-    work_items = _prep_work_items(field)
+def _translate_chars_in_list(field, translation):
+    return [_translate_chars(value, translation) for value in field]
 
-    while len(work_items) > 0:
-        work_item = work_items.pop(0)
-        oldkey = work_item[0]
-        value = work_item[1]
-        work_field = work_item[2]
-        newkey = oldkey
 
-        for t_k, t_v in six.iteritems(translation):
-            if t_k in newkey:
-                newkey = newkey.replace(t_k, t_v)
+def _translate_chars_in_key(key, translation):
+    for k, v in [(k, v) for k, v in six.iteritems(translation) if k in key]:
```

Comment:
If we're worried about performance, i think this might be able to be improved.

Currently the `[(k, v) for k, v in six.iteritems(translation) if k in key]` iterates over `key` either 1x (for escape) or 3x for unescape. Then, for every character it iterates over `key` again, so potentially another 1x or 3x times. `O(n^3)`. Also, every time a replace is done a new copy of a `key` is made.

Seems like we could do some things to make this better?
* Google lead me here: https://stackoverflow.com/questions/10017147/removing-a-list-of-characters-in-string
* Also, maybe if we simply changed `for k, v in [(k, v) for k, v in six.iteritems(translation) if k in key]:` to `for k, v in six.iteritems(translation):` that might help?

Reply:
@nmaludy i'm not sure if you missed the if condition for the inner loop. Anyway, went back to the original use of the for loop and if.

```
def _translate_chars_in_key(key, translation):
    for k, v in six.iteritems(translation):
        if k in key:
            key = key.replace(k, v)

    return key
```

Reply:
The original method will be definitely worst.  I consolidated two calls of _translate_chars into one in the unescape (and also fixed a bug there).  The list/dict comprehension is going to be less logic on deleting keys and append new dict entry.  A ran a number of orquesta integration tests between master and this branch and I am seeing completion time reduced with the change.

Reply:
Also, the string.translate method doesn't work here because the translation is a 1:1 character translation.  The translation here is 1:many or many:1 characters.

Reply:
See my benchmarks [here](https://github.com/StackStorm/st2/pull/4894#issuecomment-603580440).

This PR (now) implements the fastest versions of the options presented so far (eg: the explicit `for` loop and explicit switch on `field` type). Resolving.